winrm/client.go

```diff
@@ -23,21 +29,38 @@ type Client struct {
 
 // NewClient will create a new remote client on url, connecting with user and password
 // This function doesn't connect (connection happens only when CreateShell is called)
-func NewClient(endpoint *Endpoint, user, password string) (client *Client, err error) {
+func NewClient(endpoint *Endpoint, user, password string, auth AuthType) (client *Client, err error) {
```

Comment:
Yup, that sounds great IMO. Certificate first, then basic auth. Although, I am not sure if we should automatically retry the command using basic auth if certificate auth fails.

Rather, if both user/pass and cert is supplied, prefer certificate. The developer using this library should still know if authentication via certificate failed. Otherwise it might hide away problems the developer may not be aware of.

I for one would allow the user of the library to decide whether or not to retry with a different method.